File: packages_ts/archivist/src/query/GellishToCypherConverter.ts
import { Injectable, Logger } from '@nestjs/common';
import { Fact } from '@relica/types';
import { CacheService } from '../cache/cache.service';

interface TempEntity {
  varName: string;
  constraints: string[];
}

@Injectable()
export class GellishToCypherConverter {
  private tempEntities: Map<string, TempEntity> = new Map();
  private readonly logger = new Logger(GellishToCypherConverter.name);

  constructor(private readonly cacheService: CacheService) {}

  async processGellishQuery(
    gellishStatements: Fact[],
    page: number = 1,
    pageSize: number = 10,
  ): Promise<{ query: string; params: any }> {
    this.tempEntities.clear();
    const cypherPatterns: string[] = [];
    const params: any = {};
    const whereClauses: string[] = [];

    for (let i = 0; i < gellishStatements.length; i++) {
      const { pattern, stmtParams, whereClause } =
        await this.generateCypherPattern(gellishStatements[i], i);
      cypherPatterns.push(pattern);
      Object.assign(params, stmtParams);
      if (whereClause) whereClauses.push(whereClause);
    }

    this.logger.debug('Cypher Patterns:\n' + cypherPatterns.join('\n'));

    const skip = (page - 1) * pageSize;
    const limit = pageSize;

    const query = this.generateCypherQuery(
      cypherPatterns,
      whereClauses,
      gellishStatements.length,
      skip,
      limit,
    );

    this.logger.debug('Generated Cypher Query:\n' + query);
    this.logger.debug('Query Params:', params);

    return { query, params };
  }

  private async generateCypherPattern(
    statement: Fact,
    index: number,
  ): Promise<{ pattern: string; stmtParams: any; whereClause: string }> {
    const lhVar = this.getEntityVariable(statement.lh_object_uid);
    const rhVar = this.getEntityVariable(statement.rh_object_uid);
    const fVar = `f${index}`;
    const stmtParams: any = {};
    const whereClauses: string[] = [];

    const lhIsTemp = this.isTempUID(statement.lh_object_uid);
    const rhIsTemp = this.isTempUID(statement.rh_object_uid);
    const relIsTemp = this.isTempUID(statement.rel_type_uid);

    if (!lhIsTemp) {
      const lhDescendants = await this.cacheService.allDescendantsOf(
        statement.lh_object_uid,
      );
      stmtParams[`${lhVar}Descendants`] = [
        statement.lh_object_uid,
        ...lhDescendants,
      ];
      whereClauses.push(`${lhVar}.uid IN $${lhVar}Descendants`);
    }

    if (!rhIsTemp) {
      const rhDescendants = await this.cacheService.allDescendantsOf(
        statement.rh_object_uid,
      );
      stmtParams[`${rhVar}Descendants`] = [
        statement.rh_object_uid,
        ...rhDescendants,
      ];
      whereClauses.push(`${rhVar}.uid IN $${rhVar}Descendants`);
    }

    if (!relIsTemp) {
      const relDescendants = await this.cacheService.allDescendantsOf(
        statement.rel_type_uid,
      );
      stmtParams[`${fVar}Descendants`] = [
        statement.rel_type_uid,
        ...relDescendants,
      ];
      whereClauses.push(`${fVar}.rel_type_uid IN $${fVar}Descendants`);
    }

    const pattern = `MATCH (${lhVar}:Entity)-->(${fVar}:Fact)-->(${rhVar}:Entity)`;

    return { pattern, stmtParams, whereClause: whereClauses.join(' AND ') };
  }

  private getEntityVariable(uid: number): string {
    const varName = `var_${uid}`;
    if (!this.tempEntities.has(varName)) {
      this.tempEntities.set(varName, { varName, constraints: [] });
    }
    return varName;
  }

  private generateCypherQuery(
    patterns: string[],
    whereClauses: string[],
    statementCount: number,
    skip: number,
    limit: number,
  ): string {
    const matchClauses = patterns.join('\n');
    const whereClause =
      whereClauses.length > 0 ? `WHERE ${whereClauses.join(' AND ')}` : '';
    const returnClause = this.generateReturnClause(statementCount);

    return `
    ${matchClauses}
    ${whereClause}
    ${returnClause}
    SKIP ${skip}
    LIMIT ${limit}
  `;
  }

  private generateReturnClause(statementCount: number): string {
    const returnItems = Array.from(this.tempEntities.values()).map(
      (entity) => `${entity.varName} {.*, uid: ${entity.varName}.uid}`,
    );
    const factItems = Array(statementCount)
      .fill(0)
      .map((_, i) => `f${i}`);
    return `RETURN DISTINCT ${returnItems.join(', ')}, ${factItems.join(', ')}`;
  }

  private isTempUID(uid: number): boolean {
    return uid >= 1 && uid <= 99;
  }
}

File: packages_clj/archivist/src/io/relica/archivist/query/gellish_to_cypher_converter.clj
(ns io.relica.archivist.query.gellish-to-cypher-converter
  (:require [clojure.tools.logging :as log]
            [io.relica.common.services.cache-service :as cache :refer [cache-service-comp]]))

(defrecord TempEntity [var-name constraints])

(defn- is-temp-uid? [uid]
  (and (>= uid 1) (<= uid 99)))

(defn- get-entity-variable [temp-entities uid]
  (let [var-name (str "var_" uid)]
    (if-not (contains? @temp-entities var-name)
      (swap! temp-entities assoc var-name (->TempEntity var-name [])))
    var-name))

(defn- generate-return-clause [temp-entities statement-count]
  (println "Generating return clause with statement count:" statement-count)
  (println "Temp entities:" @temp-entities)
  (let [return-items (map (fn [[_ entity]]
                            (str (:var-name entity) " {.*, uid: " (:var-name entity) ".uid}"))
                          @temp-entities)
        fact-items (map #(str "f" %) (range statement-count))]
    (str "RETURN DISTINCT "
         (clojure.string/join ", " (concat return-items fact-items)))))

(defn- generate-cypher-query [patterns where-clauses temp-entities statement-count skip limit]
  (let [match-clauses (clojure.string/join "\n" patterns)
        where-clause (if (seq where-clauses)
                       (str "WHERE " (clojure.string/join " AND " where-clauses))
                       "")
        return-clause (generate-return-clause temp-entities statement-count)]
    (str match-clauses "\n"
         where-clause "\n"
         return-clause "\n"
         "SKIP " skip "\n"
         "LIMIT " limit)))

(defn- generate-cypher-pattern [temp-entities statement index]
  (let [lh-uid (:lh_object_uid statement)
        rh-uid (:rh_object_uid statement)
        rel-uid (:rel_type_uid statement)
        lh-var-name (get-entity-variable temp-entities lh-uid)
        rh-var-name (get-entity-variable temp-entities rh-uid)
        f-var (str "f" index)
        stmt-params (atom {})
        where-clauses (atom [])

        _ (println lh-var-name ":" rh-var-name ":" f-var)

        lh-is-temp (is-temp-uid? lh-uid)
        rh-is-temp (is-temp-uid? rh-uid)
        rel-is-temp (is-temp-uid? rel-uid)]
    (when-not lh-is-temp
      (let [lh-descendants (concat
                             [lh-uid]
                             (cache/all-descendants-of @cache-service-comp lh-uid))]
        (swap! stmt-params assoc (str lh-var-name "Descendants") lh-descendants)
        (swap! where-clauses conj (str lh-var-name ".uid IN $" lh-var-name "Descendants"))))

    (when-not rh-is-temp
      (let [rh-descendants (concat
                             [rh-uid]
                             (cache/all-descendants-of @cache-service-comp rh-uid))]
        (swap! stmt-params assoc (str rh-var-name "Descendants") rh-descendants)
        (swap! where-clauses conj (str rh-var-name ".uid IN $" rh-var-name "Descendants"))))

    (when-not rel-is-temp
      (let [rel-descendants (concat
                             [rel-uid]
                             (cache/all-descendants-of @cache-service-comp rel-uid))]
        (swap! stmt-params assoc (str f-var "Descendants") rel-descendants)
        (swap! where-clauses conj (str f-var ".rel_type_uid IN $" f-var "Descendants"))))

    (let [pattern (str "MATCH (" lh-var-name ":Entity)-->(" f-var ":Fact)-->(" rh-var-name ":Entity)")
          where-clause (if (seq @where-clauses)
                         (clojure.string/join " AND " @where-clauses)
                         nil)]

      {:pattern pattern
       :stmt-params @stmt-params
       :where-clause where-clause})))


(defn process-gellish-query
  ""
  [gellish-statements page page-size]
  (let [temp-entities (atom {})
        cypher-patterns (atom [])
        params (atom {})
        where-clauses (atom [])]

    ;; Process each statement
    (doseq [[i statement] (map-indexed vector gellish-statements)]
      (let [{:keys [pattern stmt-params where-clause]}
            (generate-cypher-pattern temp-entities statement i)]

        (swap! cypher-patterns conj pattern)
        (swap! params merge stmt-params)
        (when where-clause
          (swap! where-clauses conj where-clause))))

    (log/debug "Cypher Patterns:\n" (clojure.string/join "\n" @cypher-patterns))

    (let [skip (* (- page 1) page-size)
          limit page-size

          query (generate-cypher-query
                  @cypher-patterns
                  @where-clauses
                  temp-entities
                  (count gellish-statements)
                  skip
                  limit)]

      (log/debug "Generated Cypher Query:\n" query)
      (log/debug "Query Params:" @params)

      {:query query :params @params})))
